### PR TITLE
Add Helm chart for installing Kubeapps

### LINKED ISSUE #325: Make a Helm Chart for Kubeapps
I think many people would actually want to manage their Kubeapps installation using... Helm itself.
Today, "kubeapps up" seems too intrusive to me and having a way to track/upgrade all (most?) resources installed by kubeapps would be a nice feature, or even easily changing the installed resources (ingress, etc). Fortunately, Helm has already all we need to write a Chart!

What do you think?

### PR DESCRIPTION
**Pull request by prydonius**
fixes #325 

Todos:
- [x] values for configuring TLS with tiller-proxy

### CODE REVIEW

**Comment by prydonius** [comment]
@arapulido the latest changes should be testable, you'll need to install Tiller using `helm init` and build your own images for the dashboard and apprepository-controller

**Comment by arapulido** [comment]
@prydonius I have checked this and I am getting a CRD validation error:

```
helm install --set apprepository.image.repository=arapulido/apprepository-controller --set dashboard.image.repository=arapulido/dashboard ./kubeapps
Error: apiVersion "kubeapps.com/v1alpha1" in kubeapps/templates/apprepositories.yaml is not available
```
I have been checking around but it seems that there is not yet a solution for this. Did you encounter this error? How did you fix it?

**Comment by prydonius** [comment]
Oh woops, I forgot to mention you also need the RC Helm version: https://github.com/helm/helm/releases/tag/v2.10.0-rc.1, it adds support for a CRD hook. I was hoping for this to be released this week, but if not it will be a blocker for our release or we will need a workaround.

File: chart/kubeapps/templates/NOTES.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  export NODE_PORT=$(kubectl get --namespace {{ .Release.Namespace }} -o jsonpath="{.spec.ports[0].nodePort}" services {{ template "kubeapps.fullname" . }})
+  export NODE_IP=$(kubectl get nodes --namespace {{ .Release.Namespace }} -o jsonpath="{.items[0].status.addresses[0].address}")
+  echo http://$NODE_IP:$NODE_PORT
+{{- else if contains "LoadBalancer" .Values.frontend.service.type }}
+     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+           You can watch the status of by running 'kubectl get svc -w {{ template "kubeapps.fullname" . }}'
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "kubeapps.fullname" . }} -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+  echo http://$SERVICE_IP:{{ .Values.frontend.service.port }}
+{{- else if contains "ClusterIP" .Values.frontend.service.type }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "kubeapps.name" . }},release={{ .Release.Name }}" -o jsonpath="{.items[0].metadata.name}")
```
Comment (arapulido) [inline chart/kubeapps/templates/NOTES.txt]: The pod get the app label as .Release.Name-.kubeapps.name, instead of just kubeapps, so the command shown here will get 0 pods

File: chart/kubeapps/templates/NOTES.txt [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  echo "Visit http://127.0.0.1:8080 to use your application"
+  kubectl port-forward $POD_NAME 8080:80
```
Comment (arapulido) [inline chart/kubeapps/templates/NOTES.txt]: The pod is listening in the port 8080, so the command above should be `kubectl port-forward $POD_NAME 8080`

**Comment by arapulido** [comment]
App Repositories view is empty, even though the charts are loading fine:
screen shot 2018-07-24 at 12 27 06 pm

There are no errors in the console or the logs

**Comment by arapulido** [comment]
Installing charts doesn't work for me either

**Comment by arapulido** [comment]
@tompizmor @javsalgar It would be very useful if you start reviewing this as well, thanks!

**Comment by prydonius** [comment]
Hmm, thanks @arapulido. Seems to be an issue when building the docker image for the Dashboard, because with Telepresence the AppRepositories work for me.

With the image you built, I can see that the namespace it retrieves from the config is `undefined`: `http://localhost:8080/api/kube/apis/kubeapps.com/v1alpha1/namespaces/undefined/apprepositories`. This is also why installing charts doesn't work, because it tries to fetch the AppRepository but the namespace is not being found. Will continue looking into it.

**Comment by prydonius** [comment]
The issue was the ConfigMap was incorrect JSON (oops!), when using telepresence it uses the local config.json which was correct.

Also the creation error for AppRepositories was due to using the wrong resource link, fixed that too.

Note, when doing a Helm upgrade, please make sure you set the same password for MongoDB, otherwise it will get regenerated (known Helm issue - https://github.com/helm/helm/issues/3053) and break chart syncs and the chartsvc:

```
$ kubectl get secret $RELEASE_NAME-mongodb -o jsonpath="{.data.mongodb-root-password}" | base64 --decode
$ helm upgrade $RELEASE_NAME --set apprepository.image.repository=arapulido/apprepository-controller --set dashboard.image.repository=arapulido/dashboard --set mongodb.mongodbRootPassword=$MONGO_PASSWORD ./kubeapps
```

File: chart/kubeapps/templates/NOTES.txt (hunk 1 above)
Comment (prydonius) [reply chart/kubeapps/templates/NOTES.txt]: fixed, thanks

File: chart/kubeapps/templates/NOTES.txt (hunk 2 above)
Comment (prydonius) [reply chart/kubeapps/templates/NOTES.txt]: fixed, thanks

**Comment by arapulido** [comment]
@prydonius I have rebuilt and deployed the chart in a new cluster, and now the Dashboard doesn't get the login form, but tries directly to get to the App list (getting a 401 in response)

**Comment by arapulido** [comment]
@prydonius I have tested the chart and with the cluster-admin role things seem to work, but as soon as I create a serviceaccount with less permissions, it seems that we need to create the roles manually (when with `kubeapps up` they are created automatically). It looks like all RBAC roles are missing: https://github.com/kubeapps/kubeapps/blob/master/manifests/user-roles.jsonnet

I get this error when deploying to the namespace my serviceaccount has write access to:
screen shot 2018-07-26 at 4 41 50 pm

that also talks about the `kubeapps` namespace, which is wrong in this case.

**Comment by arapulido** [comment]
Once I created a role to read apprepositories in the **`default`** namespace and applied it in a binding to the serviceaccount I was using, I was able to deploy the chart to my namespace

File: chart/kubeapps/Chart.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
```
[the remark this answers is not in the document]
Comment (prydonius) [reply chart/kubeapps/Chart.yaml]: yes, I'll do that, thanks!

File: chart/kubeapps/templates/apprepository-deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,47 @@
+apiVersion: apps/v1beta2
+kind: Deployment
+metadata:
+  name: {{ template "kubeapps.apprepository.fullname" . }}
+  labels:
+    app: {{ template "kubeapps.apprepository.fullname" . }}
```
[the remark this answers is not in the document]
Comment (prydonius) [reply chart/kubeapps/templates/apprepository-deployment.yaml]: Where are you seeing that? I thought that was from the `kubeapps` CLI, it's not in the NOTES.txt?

[the remark this answers is not in the document]
Comment (prydonius) [reply chart/kubeapps/templates/apprepository-deployment.yaml]: Oh sorry, it's actually shown in the UI, will fix that thanks. Note you'll need to build images for the apprepository-controller and dashboard to make this actually work.

File: chart/kubeapps/templates/apprepository-deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  template:
+    metadata:
+      labels:
+        app: {{ template "kubeapps.apprepository.fullname" . }}
+        release: {{ .Release.Name }}
+    spec:
+      serviceAccountName: {{ template "kubeapps.apprepository.fullname" . }}
+      containers:
+      - name: controller
+        image: "{{ .Values.apprepository.image.repository }}:{{ .Values.apprepository.image.tag }}"
```
[the remark this answers is not in the document]
Comment (prydonius) [reply chart/kubeapps/templates/apprepository-deployment.yaml]: Maybe we should do it for consistency, but I was just going off the `helm create` template. TBH I don't really like adding the registry value, because I don't think the value needs to be so fine-grained, setting `repository=myregistry.com/apprepository-controller` works fine right?

File: chart/kubeapps/templates/NOTES.txt [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  echo http://$NODE_IP:$NODE_PORT
+{{- else if contains "LoadBalancer" .Values.frontend.service.type }}
+     NOTE: It may take a few minutes for the LoadBalancer IP to be available.
+           You can watch the status of by running 'kubectl get svc -w {{ template "kubeapps.fullname" . }}'
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "kubeapps.fullname" . }} -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+  echo http://$SERVICE_IP:{{ .Values.frontend.service.port }}
+{{- else if contains "ClusterIP" .Values.frontend.service.type }}
+  export POD_NAME=$(kubectl get pods --namespace {{ .Release.Namespace }} -l "app={{ template "kubeapps.fullname" . }},release={{ .Release.Name }}" -o jsonpath="{.items[0].metadata.name}")
+  echo "Visit http://127.0.0.1:8080 to use your application"
+  kubectl port-forward $POD_NAME 8080:8080
```
Comment (andresmgot) [inline chart/kubeapps/templates/NOTES.txt]: `kubectl port-forward --namespace {{ .Release.Namespace}} $POD_NAME 8080:8080`

Comment (prydonius) [reply chart/kubeapps/templates/NOTES.txt]: ah good catch, thanks!

Comment (migmartri) [reply chart/kubeapps/templates/NOTES.txt]: Is that what technically `kubeapps dashboard` used to do?

File: cmd/apprepository-controller/main.go [hunk 7]
```diff
@@ -75,4 +78,7 @@ func init() {
 	flag.StringVar(&kubeconfig, "kubeconfig", "", "Path to a kubeconfig. Only required if out-of-cluster.")
 	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
 	flag.StringVar(&repoSyncImage, "repo-sync-image", "kubeapps/chart-repo:latest", "container repo/image to use in CronJobs")
+	flag.StringVar(&namespace, "namespace", "kubeapps", "Namespace to discover AppRepository resources")
```
Comment (migmartri) [inline cmd/apprepository-controller/main.go]: May I ask why these changes are related to the creation of a chart? Couldn't this have been done in another previous patch?

Just trying to understand the dynamics :)

Comment (prydonius) [reply cmd/apprepository-controller/main.go]: This was needed because previously apprepository-controller was hardcoded to use the `kubeapps` namespace which was fine as `kubeapps up` didn't make that configurable. The chart can be installed in any namespace though (although we could hardcode it, but bad practice), so this is needed for that to work.

**Comment by prydonius** [comment]
@migmartri good points, I don't know if Helm supports directories inside templates, but another option is to just have the subcharts inside the `charts/` dir separating them as you suggest. I thought about it, but figured for now it wasn't that important and probably adds more boilerplate. Simply adding directories under templates/ though might be nice to clean it up a bit more.

**Comment by prydonius** [comment]
https://github.com/kubeapps/kubeapps/pull/349#issuecomment-408122586
> that also talks about the kubeapps namespace, which is wrong in this case.

@arapulido I will fix this in a separate PR as I need to rework the way we get the namespace in the dashboard a bit and don't want to add more to this. For now, this PR functionally works and so I think we can merge as-is.

File: docs/user/access-control.md [hunk 8]
```diff
@@ -131,19 +111,19 @@ kubectl create clusterrolebinding example-kubeapps-service-catalog-admin --clust
...
 ```
-kubectl create -n kubeapps rolebinding example-kubeapps-repositories-read --role=kubeapps-repositories-read --serviceaccount default:example
+kubectl create -n $KUBEAPPS_NAMESPACE rolebinding example-kubeapps-repositories-read --role=kubeapps-repositories-read --serviceaccount default:example
 ```
 
 #### Write access to App Repositories
 
 In order to create and refresh App Repositories in Kubeapps, apply the
-`kubeapps-repositories-write` Role in the kubeapps namespace.
+`kubeapps-repositories-write` Role in the namespace KUbeapps is installed in.
```
Comment (andresmgot) [inline docs/user/access-control.md]: s/KUbeapps/Kubeapps

**Comment by andresmgot** [comment]
For me I am okay to merge this PR as it is and doing different one to address the items missing to avoid more and a bigger PR.

File: chart/kubeapps/templates/apprepository-deployment.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (javsalgar) [reply chart/kubeapps/templates/apprepository-deployment.yaml]: In other charts the image is set using a template. This could be easier down the line if we want to change the way we set the images ("registry, repository, image" or "repository,image")

File: chart/kubeapps/templates/kubeapps-frontend-deployment.yaml [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        release: {{ .Release.Name }}
+    spec:
+      containers:
+      - name: nginx
+        image: "{{ .Values.frontend.image.repository }}:{{ .Values.frontend.image.tag }}"
+        livenessProbe:
+          httpGet:
+            path: /healthz
+            port: 8080
+          initialDelaySeconds: 60
```
Comment (javsalgar) [inline chart/kubeapps/templates/kubeapps-frontend-deployment.yaml]: In our production-ready infrastructure assets, these kind of values go in values.yaml

File: chart/kubeapps/templates/apprepository-deployment.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (tompizmor) [reply chart/kubeapps/templates/apprepository-deployment.yaml]: yes, I saw that Ara already built the images so I used the same ones

File: chart/kubeapps/templates/kubeapps-frontend-deployment.yaml (hunk 9 above)
Comment (tompizmor) [reply chart/kubeapps/templates/kubeapps-frontend-deployment.yaml]: right! We usually export health checks  settings so users can fine tune them

File: docs/user/access-control.md (hunk 8 above)
Comment (prydonius) [reply docs/user/access-control.md]: oops, thanks

File: chart/kubeapps/templates/apprepository-deployment.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (prydonius) [reply chart/kubeapps/templates/apprepository-deployment.yaml]: @javsalgar I couldn't find where the template was used, we seem to be just doing this: https://github.com/helm/charts/blob/master/stable/drupal/templates/deployment.yaml#L27. In any case, since this chart will eventually be going into bitnami/charts I am going to update it for consistency

**Comment by sameersbn** [comment]
@prydonius you seem to have added the tarball `charts/mongodb-4.0.4.tgz` to PR.

**Comment by prydonius** [comment]
@sameersbn I was thinking of vendoring it, but I'll remove it

File: chart/kubeapps/values.yaml [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+  image:
+    registry: docker.io
+    repository: kubeapps/apprepository-controller
+    tag: latest
+  # Image used to perform chart repository syncs
+  syncImage:
+    registry: docker.io
+    repository: kubeapps/chart-repo
+    tag: latest
+  jobsImage:
```
[the remark this answers is not in the document]
Comment (migmartri) [reply chart/kubeapps/values.yaml]: Do we run it even if you are using the version of helm that does not require the workaround?

[the remark this answers is not in the document]
Comment (prydonius) [reply chart/kubeapps/values.yaml]: We do run it yes, this workaround works with all versions of Helm, and I don't know if we can detect the new version easily enough to change the way we do this (also it's still not yet released).

[the remark this answers is not in the document]
Comment (prydonius) [reply chart/kubeapps/values.yaml]: @migmartri fixed

File: chart/kubeapps/templates/apprepository-jobs-bootstrap-rbac.yaml [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  - kubeapps.com
+  resources:
+  - apprepositories
+  verbs:
+  - get
+  - create
+  - delete
+---
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: RoleBinding
```
Comment (migmartri) [inline chart/kubeapps/templates/apprepository-jobs-bootstrap-rbac.yaml]: Add hook annotations.

File: chart/kubeapps/templates/dashboard-config.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "kubeapps.dashboard-config.fullname" . }}
+  labels:
+    app: {{ template "kubeapps.dashboard-config.fullname" . }}
+    chart: {{ template "kubeapps.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+data:
```
Comment (migmartri) [inline chart/kubeapps/templates/dashboard-config.yaml]: Eventually it might make sense to bake this configuration in the image.
